compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1863,23 +1863,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     return true;
                 }
             }
-            hir::ExprKind::Struct(qpath, fields, ..) => {
-                if let Res::Def(DefKind::Struct | DefKind::Variant, variant_def_id) =
+            hir::ExprKind::Struct(qpath, _fields, ..) => {
+                if let Res::Def(DefKind::Struct | DefKind::Variant, _variant_def_id) =
                     self.typeck_results.borrow().qpath_res(qpath, hir_id)
                 {
                     for param in
                         [param_to_point_at, fallback_param_to_point_at, self_param_to_point_at]
                     {
-                        if let Some(param) = param
-                            && self.point_at_field_if_possible(
-                                error,
-                                def_id,
-                                param,
-                                variant_def_id,
-                                fields,
-                            )
-                        {
+                        if let Some(_param) = param {
+                            // TODO: This might have broken things, since we now always return true.
+                            return false;
+                            /*
+                            let refined_expr = self
+                                .point_at_field_if_possible(todo!(), def_id, param, variant_def_id, fields)
+                                .unwrap_or_else(|expr| expr);
+                            error.obligation.cause.span = refined_expr
+                                .span
+                                .find_ancestor_in_same_ctxt(error.obligation.cause.span)
+                                .unwrap_or(refined_expr.span);
                             return true;
+                             */
```

Comment:
The old way of doing things and the new way of doing things introduced in this PR are in conflict. The old way mutate the span in place, and returns `true` if it does this.

The new way uses `Result<Expr, Expr>` to indicate whether it successfully refined or unsuccessfully refined. But it can still make a partial refinement in the unsuccessful case. The downside is that it needs to know the "root" `expr` to return it in failure.

---

I think a hybrid approach is actually going to be nicer. I will introduce a new `ErrorSpanRefiner` which can be called with:

```rs
struct FullyRefined;
struct NotFullyRefined;
fn refine_span_to_expr(&mut self, expr: &hir::Expr<'tcx>) -> Result<(), NotFullyRefined>
```

and the new/existing `point_to_*_` functions will be updated to return `Result<FullyRefined, NotFullyRefined>` instead of `bool` or `Result<&Expr, &Expr>`.

The `ErrorSpanRefiner` can also deal with updating the cause, as needed, so that the cause and the span stay in sync.